tool/tctl/common/auth_command_test.go

```diff
@@ -486,3 +499,97 @@ func TestGenerateDatabaseKeys(t *testing.T) {
 		})
 	}
 }
+
+// TestGenerateAppCertificates verifies cert/key pair generation for applications.
+func TestGenerateAppCertificates(t *testing.T) {
+	const appName = "app-1"
+	const clusterNameStr = "example.com"
+	const publicAddr = "https://app-1.example.com"
+	const sessionID = "foobar"
+
+	clusterName, err := services.NewClusterNameWithRandomID(
+		types.ClusterNameSpecV2{
+			ClusterName: clusterNameStr,
+		})
+	require.NoError(t, err)
+
+	authClient := &mockClient{
+		clusterName: clusterName,
+		userCerts: &proto.Certs{
+			SSH: []byte("SSH cert"),
+			TLS: []byte("TLS cert"),
+		},
+		appServices: []types.AppServer{
+			&types.AppServerV3{
+				Metadata: types.Metadata{
+					Name: appName,
+				},
+				Spec: types.AppServerSpecV3{
+					App: &types.AppV3{
+						Spec: types.AppSpecV3{
+							PublicAddr: publicAddr,
+						},
+					},
+				},
+			},
+		},
+		appSession: &types.WebSessionV2{
+			Metadata: types.Metadata{
+				Name: sessionID,
+			},
+		},
+	}
+
+	tests := []struct {
+		name        string
+		outDir      string
+		outFileBase string
+		appName     string
+		assertErr   require.ErrorAssertionFunc
+	}{
+		{
+			name:        "app happy path",
+			outDir:      t.TempDir(),
+			outFileBase: "app-1",
+			appName:     "app-1",
+			assertErr:   require.NoError,
+		},
+		{
+			name:        "app non-existent",
+			outDir:      t.TempDir(),
+			outFileBase: "app-2",
+			appName:     "app-2",
+			assertErr: func(t require.TestingT, err error, _ ...interface{}) {
+				require.Error(t, err)
+				require.True(t, trace.IsNotFound(err))
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			output := filepath.Join(tc.outDir, tc.outFileBase)
+			ac := AuthCommand{
+				output:        output,
+				outputFormat:  identityfile.FormatTLS,
+				signOverwrite: true,
+				genTTL:        time.Hour,
+				appName:       tc.appName,
+			}
+			err = ac.generateUserKeys(authClient)
+			tc.assertErr(t, err)
+			if err != nil {
+				return
+			}
+			require.Equal(t, proto.UserCertsRequest_App, authClient.userCertsReq.Usage)
+			require.Equal(t, tc.appName, authClient.userCertsReq.RouteToApp.Name)
+			require.Equal(t, sessionID, authClient.userCertsReq.RouteToApp.SessionID)
+			require.Equal(t, publicAddr, authClient.userCertsReq.RouteToApp.PublicAddr)
+			require.Equal(t, clusterNameStr, authClient.userCertsReq.RouteToApp.ClusterName)
```

Comment:
Can you compare the whole struct instead of individual fields? The test would be more robust this way.